### PR TITLE
gh-98458: unittest: bugfix for infinite loop while handling chained exceptions that contain cycles

### LINKED ISSUE #98458: Unittest: self-referencing explicit exception cause results in infinite loop
# Bug report

If an exception is raised with a self-referencing \_\_cause__ or \_\_context__ then TestResult._clean_tracebacks() in result.py enters an infinite loop.

Minimal example 1:
```python
try:
    raise Exception()
except Exception as e:
    raise e from e
```

Minimal example 2:
```python
try:
    e = Exception()
    raise e from e
except Exception as e:
    raise e
```

# Identified Cause
Self-references are not checked while unwinding the chained exception [on line 216 of result.py](https://github.com/python/cpython/blob/3.9/Lib/unittest/result.py#L216)

# Your environment

- CPython versions tested on: 3.9, 3.10
- Operating system and architecture: macOS

### Linked PRs
* gh-98459
* gh-99995
* gh-99996

### PR DESCRIPTION
**Pull request by AlexTate**
Closes #98458 

* Issue: gh-98458

### CODE REVIEW

**Review by zware — COMMENTED** [review]
Can you add a test for this?  It would be good to also test the more exotic case of a loop of mutually-chained exceptions, such as:

```python
try:
    raise ValueError(1)
except ValueError as e:
    try:
        raise KeyError(2) from e
    except KeyError as e2:
        raise e from e2
```

File: Lib/unittest/result.py [hunk 1]
```diff
@@ -214,7 +214,7 @@ def _clean_tracebacks(self, exctype, value, tb, test):
 
             if value is not None:
                 for c in (value.__cause__, value.__context__):
-                    if c is not None:
```
Comment (iritkatriel) [inline Lib/unittest/result.py:217]: This doesn’t help with cycles involving more than one exception. The traceback module uses a ‘seen’ set to detect cycles.

Comment (iritkatriel) [reply Lib/unittest/result.py:217]: Also, a bug fix needs to be committed with a the unit test that would have detected the bug.

**Review by kumaraditya303 — DISMISSED** [review]
Please add a test for this and address @iritkatriel's review.

**Comment by AlexTate** [comment]
Thank you for the reminder and for your patience, everyone. I've done as you have requested. Please let me know if there is anything I missed or if you believe there is a better approach

**Comment by AlexTate** [comment]
I have made the requested changes; please review again

File: Lib/test/test_unittest/test_result.py [hunk 2]
```diff
@@ -275,6 +275,39 @@ def get_exc_info():
...
+        class Foo(unittest.TestCase):
+            def test_1(self):
+                pass
+
+        def get_exc_info():
+            try:
+                # Create two directionally opposed cycles
+                # __cause__ in one direction, __context__ in the other
+                # Only one is needed since they serve the same role in _clean_tracebacks()
+                # But this way we cover both possibilities if the code changes
```
Comment (iritkatriel) [inline Lib/test/test_unittest/test_result.py]:
```suggestion
```

(not necessary)

File: Lib/test/test_unittest/test_result.py [hunk 3]
```diff
@@ -275,6 +275,39 @@ def get_exc_info():
         self.assertEqual(len(dropped), 1)
         self.assertIn("raise self.failureException(msg)", dropped[0])
 
+    def test_addFailure_filter_traceback_frames_chained_exception_cycle(self):
+        class Foo(unittest.TestCase):
+            def test_1(self):
+                pass
+
+        def get_exc_info():
```
Comment (iritkatriel) [inline Lib/test/test_unittest/test_result.py:308]: I think we should also have a test for the edge case of a single exception in a loop with itself.

File: Lib/unittest/result.py [hunk 4]
```diff
@@ -196,6 +196,7 @@ def _clean_tracebacks(self, exctype, value, tb, test):
         ret = None
         first = True
         excs = [(exctype, value, tb)]
+        seen = {id(value)}  # Handle loops in __cause__ or __context__.
```
Comment (iritkatriel) [inline Lib/unittest/result.py]:
```suggestion
        seen = {id(value)}  # Detect loops in in the chained exceptions.
```

**Review by iritkatriel — COMMENTED** [review]
Looks good, I'd add another test.

File: Lib/test/test_unittest/test_result.py (hunk 2 above)
Comment (AlexTate) [reply Lib/test/test_unittest/test_result.py]: @iritkatriel thank you for your feedback. Just to clarify, are you saying that the comment itself isn't necessary or are you saying that the approach isn't necessary?

Comment (iritkatriel) [reply Lib/test/test_unittest/test_result.py]: The comment.

Comment (AlexTate) [reply Lib/test/test_unittest/test_result.py]: Addressed in 8388d82

File: Lib/test/test_unittest/test_result.py (hunk 3 above)
Comment (AlexTate) [reply Lib/test/test_unittest/test_result.py:308]: Addressed in 6b868df

File: Lib/unittest/result.py (hunk 4 above)
Comment (AlexTate) [reply Lib/unittest/result.py]: Addressed in 08fd984

File: Lib/unittest/result.py (hunk 1 above)
Comment (AlexTate) [reply Lib/unittest/result.py:217]: Addressed in c100409

**Review by iritkatriel — APPROVED** [review]
LGTM.

@gpshead - as a maintainer of unittest, do you want to backport this?

**Review by gpshead — APPROVED** [review]

**Pull request merged**